dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go

```diff
@@ -0,0 +1,385 @@
+// Need to consider confliction between rules.
+func handleBucketLifecyle(bucket string, defaultBackend string, rules []*osdss3.LifecycleRule) error {
+	// Translate rules set by user to internal rules which can be sorted.
+
+	// Lifecycle scheduling must be mutual excluded among several schedulers, so get lock first.
+	ret := db.DbAdapter.LockBucketLifecycleSched(bucket)
+	for i := 0; i < 3; i++ {
+		if ret == LockSuccess {
+			break
+		} else if ret == LockBusy {
+			return fmt.Errorf("Lifecycle scheduleing of bucket[%s] is in progress.")
+		} else {
+			// Try to lock again, try three times at most
+			ret = db.DbAdapter.LockBucketLifecycleSched(bucket)
+		}
+	}
+	if ret != LockSuccess {
+		log.Logf("Lock scheduling failed.\n")
+		return fmt.Errorf("Internal error: Lock scheduling failed.")
+	}
+	// Make sure unlock before return
+	defer db.DbAdapter.UnlockBucketLifecycleSched(bucket)
+
+	var inRules, abortRules InterRules
+	for _, rule := range rules {
+		if rule.Status == RuleStatusDisabled {
+			continue
+		}
+
+		// actions
+		for _, ac := range rule.Actions {
+			if ac.Backend == "" {
+				// if no backend specified, then use the default backend of the bucket
+				ac.Backend = defaultBackend
+			}
+			var acType int
+			if ac.Name == ActionNameExpiration {
+				// Expiration
+				acType = ActionExpiration
+			} else if ac.Backend == defaultBackend {
+				acType = ActionIncloudTransition
+			} else {
+				acType = ActionCrosscloudTransition
+			}
+
+			v := InternalLifecycleRule{Bucket: bucket, Days: ac.GetDays(), ActionType: acType, Tier: ac.GetTier(), Backend: ac.GetBackend()}
+			if rule.GetFilter() != nil {
+				v.Filter = InternalLifecycleFilter{Prefix: rule.Filter.Prefix}
+			}
+			inRules = append(inRules, &v)
+		}
+
+		if rule.AbortIncompleteMultipartUpload.DaysAfterInitiation > 0 {
+			// abort incomplete multipart uploads
+			abortRule := InternalLifecycleRule{Bucket: bucket, Days: rule.AbortIncompleteMultipartUpload.DaysAfterInitiation, ActionType: AbortIncompleteMultipartUpload}
+			if rule.GetFilter() != nil {
+				abortRule.Filter = InternalLifecycleFilter{Prefix: rule.Filter.Prefix}
+			}
+			abortRules = append(abortRules, &abortRule)
+		}
+	}
+
+	// Sort rules, in case different actions exist for an object at the same time, for example, expiration after 30 days
+	// and transition to azure after 60 days, we need to make sure only one action will be taken, and that needs the
+	// sorting be stable.
+	sort.Stable(inRules)
+	// Begin: Log for debug
+	for _, v := range inRules {
+		log.Logf("action rule: %+v\n", *v)
+	}
+	// End: Log for debug
+	schedSortedActionsRules(&inRules)
+
+	sort.Stable(abortRules)
+	// Begin: Log for debug
+	for _, v := range abortRules {
+		log.Logf("abort rule: %+v\n", *v)
+	}
+	// End: Log for debug
+	schedSortedAbortRules(&abortRules)
+
+	return nil
+}
+
+func checkTransitionValidation(source int32, destination int32) bool {
+	valid := true
```

Comment:
There is no need to define valid variable here, just return true or false.

Reply:
done